go_pprof_scraper/datadog_checks/go_pprof_scraper/check.py

```diff
@@ -0,0 +1,208 @@
+            # self.duration so that all of the returned profiles reflect the
+            # change in program state over the same time period
+            time.sleep(self.duration)
+        else:
+            query_params["seconds"] = self.duration
+        url = urljoin(self.url, "profile" if profile == "cpu" else profile)
+
+        response = self.http.get(
+            url,
+            params=query_params,
+            # We want to stream the response so that we can forward it to the
+            # upload, ideally without having to actually read the whole profile
+            # into memory
+            stream=True,
+            # We can get bit by the default timeout here when passing the "seconds"
+            # query parameter. By design, the request takes that long to finish.
+            #
+            # requests supports a timeout parameter, which can be a tuple
+            # (connect_timeout, read_timeout). The connect_timeout can be short, but
+            # the read_timeout needs to be longer than the profile duration.
+            timeout=(3.05, self.duration + 3),
+        )
+
+        response.raise_for_status()
+        if profile == "cpu" or self.cumulative:
+            name = profile
+        else:
+            name = "delta-{}".format(profile)
+        return name, response
+
+    def check(self, _):
+        if not self.apm_enabled:
+            self.service_check("can_connect", AgentCheck.CRITICAL, message="the trace agent is not enabled")
+            return
+
+        try:
+            start = datetime.datetime.utcnow()
+            with ThreadPoolExecutor() as executor:
+                profiles = list(executor.map(self._get_profile, self.profiles))
+
+            files = [
+                # We want response.raw, because we're going to take the response
+                # and feed it straight into the profile upload (see
+                # self._get_profile)
+                ("data[%s.pprof]" % name, ("pprof-data", response.raw, "application/octet-stream"))
+                for name, response in profiles
+            ]
+
+            def add_form_field(name, value):
+                files.append((name, (None, value)))
+
+            add_form_field("host", self.hostname)
+            add_form_field("version", "3")
+            add_form_field("format", "pprof")
+            add_form_field("family", "go")
+            add_form_field("start", _timeformat(start))
+            add_form_field("end", _timeformat(start + datetime.timedelta(seconds=self.duration)))
+            add_form_field("tags[]", "runtime:go")
+            add_form_field("tags[]", "service:{}".format(self.service))
+            add_form_field("tags[]", "runtime-id:{}".format(self.runtime_id))
+            if self.env:
+                add_form_field("tags[]", "env:{}".format(self.env))
+            for tag in self.tags:
+                add_form_field("tags[]", tag)
+
+            # TODO: container ID? If we can get it, it should be added as a
```

Comment:
Is there a Python API for getting other tags associated with the service being checked? Specifically if this is running on a cluster using autodiscovery, and if there's a container ID tag available.

Reply:
I asked the team and it looks like this is not possible

Reply:
Ok, thanks for looking into it. It's good to have this info, but not critical enough to block this feature. I can investigate further later.